File: backend/socket/rate-limit.ts
import { isRelease } from '@/util/version.util';
import { GAME_VERSION, RATE_LIMIT_CONFIG } from '@/constant/game.constant';

/** Rate limiting is bypassed entirely outside a release build. */
export const skipIfDev = (): boolean => !isRelease(GAME_VERSION);

export interface RateLimiter {
    name: string;
    consume(key: string): { allowed: true } | { allowed: false; retryAfterMs: number };
}

/** In-memory sliding window, no external dependency. */
export function createSlidingWindow(name: string, cfg: { windowMs: number; limit: number }): RateLimiter {
    const hits = new Map<string, number[]>();
    const fresh = (timestamps: number[], now: number) => timestamps.filter(t => now - t < cfg.windowMs);

    const gc = setInterval(() => {
        const now = Date.now();
        for (const [key, timestamps] of hits.entries()) {
            const kept = fresh(timestamps, now);
            if (kept.length === 0)
                hits.delete(key);
            else if (kept.length !== timestamps.length)
                hits.set(key, kept);
        }
    }, cfg.windowMs);
    gc.unref();

    return {
        name,
        consume(key: string) {
            if (skipIfDev())
                return { allowed: true };

            const now = Date.now();
            const timestamps = fresh(hits.get(key) ?? [], now);

            if (timestamps.length >= cfg.limit)
                return { allowed: false, retryAfterMs: timestamps[0] + cfg.windowMs - now };

            timestamps.push(now);
            hits.set(key, timestamps);

            return { allowed: true };
        },
    };
}

export const battleLimiter = createSlidingWindow('battle', RATE_LIMIT_CONFIG.battle);
export const shopLimiter = createSlidingWindow('shop', RATE_LIMIT_CONFIG.shop);
export const floodLimiter = createSlidingWindow('flood', RATE_LIMIT_CONFIG.flood);

File: lib/mini_lineage/game/rate_limit.ex
defmodule MiniLineage.Game.RateLimit do
  @moduledoc """
  In-memory sliding window, no external dependency. Bypassed entirely unless enabled, so local
  development is never throttled — matching the reference, which keyed that off a release build.
  """
  use GenServer

  @table :mini_lineage_rate_limit

  @limits %{
    battle: %{window_ms: 60_000, limit: 60},
    shop: %{window_ms: 60_000, limit: 30},
    flood: %{window_ms: 60_000, limit: 300}
  }

  def start_link(_opts), do: GenServer.start_link(__MODULE__, :ok, name: __MODULE__)

  @impl true
  def init(:ok) do
    :ets.new(@table, [:named_table, :public, :duplicate_bag, read_concurrency: true])
    schedule_sweep()

    {:ok, %{}}
  end

  @impl true
  def handle_info(:sweep, state) do
    now = System.monotonic_time(:millisecond)
    widest = @limits |> Map.values() |> Enum.map(& &1.window_ms) |> Enum.max()

    :ets.select_delete(@table, [{{:_, :"$1"}, [{:<, :"$1", now - widest}], [true]}])
    schedule_sweep()

    {:noreply, state}
  end

  defp schedule_sweep, do: Process.send_after(self(), :sweep, 60_000)

  @doc "Returns `:ok`, or `{:error, retry_after_ms}` when the window is full."
  def check(key, limiter) when is_map_key(@limits, limiter) do
    if Application.get_env(:mini_lineage, :rate_limit, false),
      do: consume(key, limiter),
      else: :ok
  end

  defp consume(key, limiter) do
    %{window_ms: window_ms, limit: limit} = @limits[limiter]
    now = System.monotonic_time(:millisecond)
    cutoff = now - window_ms

    fresh =
      @table
      |> :ets.lookup({key, limiter})
      |> Enum.map(&elem(&1, 1))
      |> Enum.filter(&(&1 > cutoff))

    if length(fresh) >= limit do
      {:error, Enum.min(fresh) + window_ms - now}
    else
      :ets.insert(@table, {{key, limiter}, now})
      :ok
    end
  end
end
